app/app.go

```diff
@@ -602,12 +613,101 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
 // EndBlocker application updates every end block
 func (app *RegenApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	resp := app.mm.EndBlock(ctx, req)
-	events, vals := app.smm.EndBlock(ctx, req)
-	if len(resp.ValidatorUpdates) > 0 && len(vals) > 0 {
-		panic("validator EndBlock updates already set by the SDK Module Manager")
+
+	// at the upgrade height
+	// for all validators in app state (app.StakingKeeper.GetAllValidators), create a ValidatorUpdate with the correct power.
+	// for all remaining validators in []VoteInfo that are not in app state, set their voting power to zero.
+	// reset jailed validator missed blocks counter so that the validator won't be immediately slashed for downtime after unjail.
+
+	// after the upgrade height continue the correct behavior.
+	if ctx.BlockHeight() == UPGRADE_HEIGHT {
```

Comment:
`UPGRADE_HEIGHT` should not be a `const`. Instead, let's add a field to `RegenApp` called `runEndBlockerUpgrade` that gets set in the upgrade handle
```suggestion
	if app.runEndBlockerUpgrade {
	   app.runEndBlockerUpgrade = false
```